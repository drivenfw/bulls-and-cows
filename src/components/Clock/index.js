import React, { Component } from 'react'
import styled from 'styled-components'

import { granularTime, handLength } from '../../helpers/clock'


const StyledClock = styled.div`
  height: 70px;
  border: 1px solid grey;
  border-radius: 50%;
  box-shadow: 0 0 5px 3px lightgrey;
  padding: 3px;
`

const SecondHand = styled.div`
  width: ${props => props.width}px;
  height: 1px;
  background:grey;
  transform-origin: 100%;
  transform: 
    translateX(${props => props.translateX}px) 
    translateY(34px) 
    rotate(${props => 90 + props.rotate}deg);
`

class Clock extends Component {
  state = { 
    rootElWidth: 1,
    rootElHeight: 1
  }

  rootEl = React.createRef()

  widthChange = () => {
    const { rootElWidth, rootElHeight } = this.state
    const { clientWidth, clientHeight } = this.rootEl.current

    if (rootElWidth !== clientWidth || rootElHeight !== clientHeight) {
      this.setState({
        rootElWidth: clientWidth,
        rootElHeight: clientHeight
      }) 
    }
  }

  componentDidMount() {
    window.addEventListener('resize', this.widthChange)
    this.setState({ 
      rootElWidth: this.rootEl.current.clientWidth,
      rootElHeight: this.rootEl.current.clientHeight
    })
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.widthChange)
  }

  render() {
    const { rootElWidth, rootElHeight } = this.state

    const time = granularTime(this.props.timeSecs)
    const angleDeg = 6 * time.secs
    const angleRad = Math.PI / 180 * angleDeg
    const width = handLength(rootElWidth, rootElHeight, angleRad)
    const translateX = rootElWidth / 2 - width - 5

    return (
      <StyledClock innerRef={this.rootEl}>
        <SecondHand 
          rotate={angleDeg} 
          width={width}
          translateX={translateX}
        />
      </StyledClock>
    )
  }
}

Clock.defaultProps = {
  timeSecs: 43
}

export default Clock

