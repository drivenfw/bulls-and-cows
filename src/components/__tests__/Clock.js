import React from 'react'
import { configure, shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import toJson from 'enzyme-to-json'

import Clock from 'components/Clock'


configure({ adapter: new Adapter() })

describe('<Clock />', () => {
  describe('snapshot test', () => {
    it('should match the snapshot', () => {
      const tree = mount(<Clock />)

      expect(toJson(tree)).toMatchSnapshot()
    })
  })
})

