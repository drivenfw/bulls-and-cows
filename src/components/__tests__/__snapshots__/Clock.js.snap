// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Clock /> snapshot test should match the snapshot 1`] = `
<Clock
  timeSecs={0}
>
  <styled.div
    innerRef={
      Object {
        "current": <div
          class="sc-htpNat bVcGTz"
      >
          <div
              class="sc-bZQynM hhTrkN"
          >
              <div
                  class="sc-EHOje deRQDs"
              >
                  <div
                      class="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                  >
                      <div>
                          <div
                              class="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                          >
                              0
                          </div>
                      </div>
                  </div>
                  <div
                      class="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                  >
                      <div>
                          <div
                              class="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                          >
                              0
                          </div>
                      </div>
                  </div>
              </div>
              <div
                  class="sc-EHOje deRQDs"
              >
                  <div
                      class="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                  >
                      <div>
                          <div
                              class="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                          >
                              0
                          </div>
                      </div>
                  </div>
                  <div
                      class="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                  >
                      <div>
                          <div
                              class="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                          >
                              0
                          </div>
                      </div>
                  </div>
              </div>
          </div>
          <div
              class="sc-gzVnrw hvMpkW"
              rotate="0"
              width="NaN"
          />
      </div>,
      }
    }
  >
    <div
      className="sc-htpNat bVcGTz"
    >
      <styled.div>
        <div
          className="sc-bZQynM hhTrkN"
        >
          <styled.div>
            <div
              className="sc-EHOje deRQDs"
            >
              <Styled(Carousel)
                direction="down"
              >
                <Carousel
                  className="sc-bxivhb dQNVgY"
                  direction="down"
                >
                  <styled.div
                    className="sc-bxivhb dQNVgY"
                  >
                    <div
                      className="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      >
                        <div>
                          <CSSTransition
                            classNames="scroll-down"
                            in={true}
                            key=".$.$0"
                            mountOnEnter={true}
                            onExited={[Function]}
                            timeout={500}
                            unmountOnExit={true}
                          >
                            <Transition
                              appear={false}
                              enter={true}
                              exit={true}
                              in={true}
                              mountOnEnter={true}
                              onEnter={[Function]}
                              onEntered={[Function]}
                              onEntering={[Function]}
                              onExit={[Function]}
                              onExited={[Function]}
                              onExiting={[Function]}
                              timeout={500}
                              unmountOnExit={true}
                            >
                              <Styled(styled.div)
                                key="0"
                              >
                                <styled.div
                                  className="sc-ifAKCX eXRWbD"
                                >
                                  <div
                                    className="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                                  >
                                    0
                                  </div>
                                </styled.div>
                              </Styled(styled.div)>
                            </Transition>
                          </CSSTransition>
                        </div>
                      </TransitionGroup>
                    </div>
                  </styled.div>
                </Carousel>
              </Styled(Carousel)>
              <Styled(Carousel)
                direction="down"
              >
                <Carousel
                  className="sc-bxivhb dQNVgY"
                  direction="down"
                >
                  <styled.div
                    className="sc-bxivhb dQNVgY"
                  >
                    <div
                      className="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      >
                        <div>
                          <CSSTransition
                            classNames="scroll-down"
                            in={true}
                            key=".$.$0"
                            mountOnEnter={true}
                            onExited={[Function]}
                            timeout={500}
                            unmountOnExit={true}
                          >
                            <Transition
                              appear={false}
                              enter={true}
                              exit={true}
                              in={true}
                              mountOnEnter={true}
                              onEnter={[Function]}
                              onEntered={[Function]}
                              onEntering={[Function]}
                              onExit={[Function]}
                              onExited={[Function]}
                              onExiting={[Function]}
                              timeout={500}
                              unmountOnExit={true}
                            >
                              <Styled(styled.div)
                                key="0"
                              >
                                <styled.div
                                  className="sc-ifAKCX eXRWbD"
                                >
                                  <div
                                    className="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                                  >
                                    0
                                  </div>
                                </styled.div>
                              </Styled(styled.div)>
                            </Transition>
                          </CSSTransition>
                        </div>
                      </TransitionGroup>
                    </div>
                  </styled.div>
                </Carousel>
              </Styled(Carousel)>
            </div>
          </styled.div>
          <styled.div>
            <div
              className="sc-EHOje deRQDs"
            >
              <Styled(Carousel)
                direction="down"
              >
                <Carousel
                  className="sc-bxivhb dQNVgY"
                  direction="down"
                >
                  <styled.div
                    className="sc-bxivhb dQNVgY"
                  >
                    <div
                      className="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      >
                        <div>
                          <CSSTransition
                            classNames="scroll-down"
                            in={true}
                            key=".$.$0"
                            mountOnEnter={true}
                            onExited={[Function]}
                            timeout={500}
                            unmountOnExit={true}
                          >
                            <Transition
                              appear={false}
                              enter={true}
                              exit={true}
                              in={true}
                              mountOnEnter={true}
                              onEnter={[Function]}
                              onEntered={[Function]}
                              onEntering={[Function]}
                              onExit={[Function]}
                              onExited={[Function]}
                              onExiting={[Function]}
                              timeout={500}
                              unmountOnExit={true}
                            >
                              <Styled(styled.div)
                                key="0"
                              >
                                <styled.div
                                  className="sc-ifAKCX eXRWbD"
                                >
                                  <div
                                    className="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                                  >
                                    0
                                  </div>
                                </styled.div>
                              </Styled(styled.div)>
                            </Transition>
                          </CSSTransition>
                        </div>
                      </TransitionGroup>
                    </div>
                  </styled.div>
                </Carousel>
              </Styled(Carousel)>
              <Styled(Carousel)
                direction="down"
              >
                <Carousel
                  className="sc-bxivhb dQNVgY"
                  direction="down"
                >
                  <styled.div
                    className="sc-bxivhb dQNVgY"
                  >
                    <div
                      className="sc-bxivhb dQNVgY sc-bdVaJa hJKpzc"
                    >
                      <TransitionGroup
                        childFactory={[Function]}
                        component="div"
                      >
                        <div>
                          <CSSTransition
                            classNames="scroll-down"
                            in={true}
                            key=".$.$0"
                            mountOnEnter={true}
                            onExited={[Function]}
                            timeout={500}
                            unmountOnExit={true}
                          >
                            <Transition
                              appear={false}
                              enter={true}
                              exit={true}
                              in={true}
                              mountOnEnter={true}
                              onEnter={[Function]}
                              onEntered={[Function]}
                              onEntering={[Function]}
                              onExit={[Function]}
                              onExited={[Function]}
                              onExiting={[Function]}
                              timeout={500}
                              unmountOnExit={true}
                            >
                              <Styled(styled.div)
                                key="0"
                              >
                                <styled.div
                                  className="sc-ifAKCX eXRWbD"
                                >
                                  <div
                                    className="sc-ifAKCX eXRWbD sc-bwzfXH jvxqEc"
                                  >
                                    0
                                  </div>
                                </styled.div>
                              </Styled(styled.div)>
                            </Transition>
                          </CSSTransition>
                        </div>
                      </TransitionGroup>
                    </div>
                  </styled.div>
                </Carousel>
              </Styled(Carousel)>
            </div>
          </styled.div>
        </div>
      </styled.div>
      <styled.div
        rotate={0}
        width={NaN}
      >
        <div
          className="sc-gzVnrw hvMpkW"
          rotate={0}
          width={NaN}
        />
      </styled.div>
    </div>
  </styled.div>
</Clock>
`;
